### ISSUE TITLE
[gkd recipe] jsd mode silently clamps beta to (1e-6, 1-1e-6): beta: 1.0 yields a ~0 loss with no warning

### ISSUE THREAD

**Issue by XuMeng00124** [verl-project/verl-recipe#151]
## Summary

`gkd/megatron/megatron_distill_losses.py` (`_VocabParallelJSDivergence`) implements the GKD-paper "coefficient-matched" generalized JSD:

```python
beta = min(max(float(beta), 1e-6), 1.0 - 1e-6)   # clamped
M = beta * P + (1 - beta) * Q                      # P = teacher (top-k), Q = student
per_token_jsd = beta * KL(P || M) + (1 - beta) * [KL(Q || M)_topk + KL(Q || M)_rest]
```

Two consequences:

1. **The loss vanishes at both endpoints of the clamp.** As β→1⁻, a first-order expansion gives `D(β) ≈ (1−β)·KL(Q‖P) → 0`; symmetrically `D(β) ≈ β·KL(P‖Q) → 0` as β→0⁺. Near the endpoints β acts as a *signal-strength* knob (≈ learning-rate scaling), not a mixture-dominance knob: at β=0.9 the gradient magnitude is only ~1/15 of pure reverse KL (finite-difference measurement on the same formula — details and production-run measurements in the cross-referenced ms-swift issue below).
2. **`beta: 1.0` is a silent no-op.** The clamp maps it to 1−1e−6, so the loss is ~2e−6 (measured on the idealized full-vocabulary case; the analytic rest term contributes `O((1−β)·log(1/(1−β)))` which also vanishes). A user selecting `name: jsd, beta: 1.0` — a natural choice meaning "teacher-dominant mixture" — gets a training run with effectively zero distillation signal and **no warning anywhere**.

The default `beta=0.5` (symmetric JSD) is unaffected — which is why this is easy to miss.

Note: the coefficient-matched form is exactly Eq. (1) of the GKD paper (arXiv:2306.13649), which itself derives `lim_{β→0} D(β)/β = KL(P‖Q)` (Huszár 2015) — the *scaling* near the endpoints is expected math, not a bug. ms-swift closed our related issue (modelscope/ms-swift#10078) as "matches paper, no change planned", and we are not asking verl to change the formula's convention here. What has no paper-fidelity defense is the verl-specific behavior: unlike ms-swift (whose exact β=0/1 branches return the full pure KLs), the verl-recipe `jsd` mode **clamps β into (1e-6, 1−1e-6), so `beta: 1.0` / `beta: 0.0` silently produce a near-zero loss with no warning**.

## Self-contained repro (numpy, mirrors the formula)

```python
import numpy as np
rng = np.random.default_rng(7)
z = rng.normal(0, 1.5, 300); w = rng.normal(0, 1.5, 300) + 0.3; w[:3] += 2.0
softmax = lambda x: np.exp(x - x.max()) / np.exp(x - x.max()).sum()
Q, P = softmax(z), softmax(w)          # student, teacher
kl = lambda a, b: np.sum(a * np.log(a / b))
RKL, FKL = kl(Q, P), kl(P, Q)          # reverse / forward KL

def jsd(beta_user):                     # == _VocabParallelJSDivergence (idealized, full vocab)
    beta = min(max(float(beta_user), 1e-6), 1.0 - 1e-6)
    M = beta * P + (1 - beta) * Q
    return beta * kl(P, M) + (1 - beta) * kl(Q, M)

for b in [0.001, 0.1, 0.5, 0.9, 0.999, 1.0]:
    print(f"beta={b:<6} jsd={jsd(b):.4f}")
# reverse KL = 2.0450, forward KL = 1.9262 on this seed
```

Output:

| β | jsd loss | vs pure KL |
|---|---|---|
| 0.001 | 0.0019 | ≈ 0.001 × forward KL |
| 0.100 | 0.1454 | ≈ 0.08 × forward KL |
| 0.500 | 0.3492 | symmetric JSD (fine) |
| 0.900 | 0.1474 | ≈ 0.07 × reverse KL |
| 0.999 | 0.0020 | ≈ 0.001 × reverse KL |
| 1.000 | **0.000002** | **clamped to 1−1e−6 → silent no-op** |

Note the loss at β=0.9 is *lower* than at β=0.5, and monotonically → 0 at both ends: there is no value of β in `jsd` mode that approaches either pure KL.

## Practical impact

We hit this in a 27B on-policy distillation setup (measured on the identical interior formula in ms-swift, see cross-reference): a run configured with β=0.9 showed loss/grad_norm at ~1/13–1/15 of an otherwise-identical β=1 run at iteration 1 — i.e. β near the endpoint effectively rescales the learning rate rather than blending divergence directions. In verl-recipe the failure mode is even easier to trigger: since the `jsd` mode clamps β, `beta: 1.0` (or 0.0) yields a near-zero loss with no signal that anything is wrong.

For comparison, the other three modes in this file are exact and endpoint-clean: `kl` (forward KL), `rkl` (reverse KL), and `kl_rkl` (a linear blend `(1−r)·KL + r·RKL` of two *full* KLs, which is continuous across its whole range by construction). Only `jsd` has the collapsing parameterization.

## Suggested fix

**(a) Primary ask — stop failing silently**: emit a warning (or raise) when `beta >= 0.99` or `beta <= 0.01` in `jsd` mode, and document that the `jsd` mode cannot approach either pure KL — users wanting a forward/reverse blend should use `kl_rkl`, and pure KLs are the `kl`/`rkl` modes.

**(b) Alternative — endpoint-continuous parameterization** (same mixture M, coefficients swapped):

```python
# current
per_token_jsd = beta * kl_P_M + one_minus_beta * (kl_Q_M_topk + kl_Q_M_rest)
# alternative
per_token_jsd = one_minus_beta * kl_P_M + beta * (kl_Q_M_topk + kl_Q_M_rest)
```

Then β=1 → exactly `KL(Q‖P)` (reverse KL), β=0 → exactly `KL(P‖Q)` (forward KL), β=0.5 numerically unchanged (symmetric JSD), and the whole [0,1] range interpolates continuously. The swapped form is still a valid divergence (non-negative, zero iff P=Q), and the clamp could then be relaxed to exact endpoints. Note: ms-swift declined this exact change (modelscope/ms-swift#10078) in favor of fidelity to the GKD paper / TRL, so we list it as an alternative — the warning in (a) is the main ask.

We filed the same root cause against ms-swift (their `jsd_loss` interior has the identical coefficients, with hardcoded pure-KL endpoint branches that make the loss jump-discontinuous instead of clamped): https://github.com/modelscope/ms-swift/issues/10078 — that issue contains the full first-order analysis, the gradient finite-difference check, and the production-run loss/grad-norm measurements.

## Environment

- verl-recipe `main` (checked 2026-09-09): `gkd/megatron/megatron_distill_losses.py`, `_VocabParallelJSDivergence.forward` (clamp at function head; `M_topk = beta * P_topk + one_minus_beta * Q_topk`; `tmp = beta * kl_P_M_local + one_minus_beta * kl_Q_M_topk_local`; rest term `one_minus_beta * kl_Q_M_rest`).
- Main repo `volcengine/verl` OPD engine (forward top-k KL / reverse KL) appears unaffected — this issue is scoped to the gkd recipe.
- Repro is pure numpy, no GPU needed.
